optimization.py

```diff
@@ -318,4 +345,243 @@ def _get_variable_name(self, param_name):
+    """See base class."""
+    assignments = []
+    # get the local step 
+    steps = tf.cast(global_step, tf.float32) + 1.
+    for (grad, param) in grads_and_vars:
+      if grad is None or param is None:
+        continue
+
+      param_name = self._get_variable_name(param.name)
+      
+      m = tf.compat.v1.get_variable(
+          name=param_name + "/nadam_m",
+          shape=param.shape.as_list(),
+          dtype=tf.float32,
+          trainable=False,
+          initializer=tf.zeros_initializer())
+      v = tf.compat.v1.get_variable(
+          name=param_name + "/nadam_v",
+          shape=param.shape.as_list(),
+          dtype=tf.float32,
+          trainable=False,
+          initializer=tf.zeros_initializer())
+
+      # Standard Adam update.
+      next_m = (tf.multiply(self.beta_1, m) + tf.multiply(1.0 - self.beta_1, grad))
+      next_v = (tf.multiply(self.beta_2, v) + tf.multiply(1.0 - self.beta_2, tf.square(grad)))
+     
+      # We could use momentum scheduling variable 
+      # mu_t   = self.beta_1 * (1. - 0.5 * (0.96**(0.004*steps))) 
+      # instead we use constant scheduling so mu_t = self_beta1       
+     
+      beta1_correction = 1./(1. - (self.beta_1 ** steps))
+      beta1_correction_tp1 = 1./(1. - (self.beta_1 ** (steps+1)))
+      beta2_correction = 1./(1. - (self.beta_2 ** steps))
+
+      next_m_unbiased = tf.multiply(beta1_correction_tp1,next_m)
+      next_v_unbiased = tf.multiply(beta2_correction,next_v)
+      # Nesterov addition moment calculation
+      
+      next_m_nesterov = (tf.multiply(self.beta_1, next_m_unbiased) + tf.multiply((1.0-self.beta_1)*beta1_correction,grad))
+
+      update = next_m_nesterov / (tf.sqrt(next_v_unbiased) + self.epsilon)
+
+      # Just adding the square of the weights to the loss function is *not*
+      # the correct way of using L2 regularization/weight decay with Adam,
+      # since that will interact with the m and v parameters in strange ways.
+      #
+      # Instead we want ot decay the weights in a manner that doesn't interact
+      # with the m/v parameters. This is equivalent to adding the square
+      # of the weights to the loss with plain (non-momentum) SGD.
+      if self._do_use_weight_decay(param_name):
+        update += self.weight_decay_rate * param
+
+      update_with_lr = self.learning_rate * update
+
+      next_param = param - update_with_lr
+
+      assignments.extend(
+        [param.assign(next_param),
+          m.assign(next_m),
+          v.assign(next_v)])
+    return tf.group(*assignments, name=name)
+
+  def _do_use_weight_decay(self, param_name):
+    """Whether to use L2 weight decay for `param_name`."""
+    if not self.weight_decay_rate:
+      return False
+    if self.exclude_from_weight_decay:
+      for r in self.exclude_from_weight_decay:
+        if re.search(r, param_name) is not None:
+          return False
+    return True
+
+  def _get_variable_name(self, param_name):
```

Comment:
looks like the same def _get_variable_name() repeated four times - maybe move to some base class?

Reply:
The idea here is to preserve the original BERT-AdamOptimizer code. This is why, I did it this way. Is more question of correlation and legibility with the original BERT code. This of course 
does not impact run-time.